hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/execution/datasources/HoodieFileGroupReaderBasedFileFormat.scala

```diff
@@ -0,0 +1,381 @@
+                fileReaders, filters, requiredFilters, storageConf, metaClient.getTableConfig)
+              val props = metaClient.getTableConfig.getProps
+              options.foreach(kv => props.setProperty(kv._1, kv._2))
+              val reader = new HoodieFileGroupReader[InternalRow](
+                readerContext,
+                new HoodieHadoopStorage(metaClient.getBasePath, storageConf),
+                tablePath,
+                queryTimestamp,
+                fileSlice,
+                dataAvroSchema,
+                requestedAvroSchema,
+                internalSchemaOpt,
+                metaClient,
+                props,
+                file.start,
+                file.length,
+                shouldUseRecordPosition)
+              reader.initRecordIterators()
+              // Append partition values to rows and project to output schema
+              appendPartitionAndProject(
+                reader.getClosableIterator,
+                requestedSchema,
+                remainingPartitionSchema,
+                outputSchema,
+                fileSliceMapping.getPartitionValues,
+                fixedPartitionIndexes)
+
+            case _ =>
+              val filePath = sparkAdapter.getSparkPartitionedFileUtils.getPathFromPartitionedFile(file)
+              val baseFileFormat = detectFileFormat(filePath.toString)
+              baseFileFormat match {
+                case PARQUET_FILE_EXTENSION => readBaseFile(file, broadcastedParquetFileReader.value, requestedSchema, remainingPartitionSchema, fixedPartitionIndexes,
+              requiredSchema, partitionSchema, outputSchema, filters, storageConf)
+                case ORC_FILE_EXTENSION => readBaseFile(file, broadcastedOrcFileReader.value, requestedSchema, remainingPartitionSchema, fixedPartitionIndexes,
+                  requiredSchema, partitionSchema, outputSchema, filters, storageConf)
+              }
+          }
+        // CDC queries.
+        case hoodiePartitionCDCFileGroupSliceMapping: HoodiePartitionCDCFileGroupMapping =>
+          buildCDCRecordIterator(hoodiePartitionCDCFileGroupSliceMapping, broadcastedParquetFileReader.value, storageConf, fileIndexProps, requiredSchema)
+
+        case _ =>
+          val filePath = sparkAdapter.getSparkPartitionedFileUtils.getPathFromPartitionedFile(file)
+          val baseFileFormat = detectFileFormat(filePath.toString)
+          baseFileFormat match {
+            case PARQUET_FILE_EXTENSION => readBaseFile(file, broadcastedParquetFileReader.value, requestedSchema, remainingPartitionSchema, fixedPartitionIndexes,
+              requiredSchema, partitionSchema, outputSchema, filters, storageConf)
+            case ORC_FILE_EXTENSION => readBaseFile(file, broadcastedOrcFileReader.value, requestedSchema, remainingPartitionSchema, fixedPartitionIndexes,
+              requiredSchema, partitionSchema, outputSchema, filters, storageConf)
+          }
+      }
+    }
```

Comment:
Need to add in CloseableIteratorListener.addListener() if we don't unify this with parquet file format